### ISSUE TITLE
Per class F1 score multiclass Unet Metrics

### ISSUE THREAD

**Issue by Abdulk084** [qubvel/segmentation_models#484]
I am training a segmentation model from the [repository][1]. I have used the following loss functions and metrics during my training process.

**loss:**

    dice_loss = sm.losses.DiceLoss(class_weights=np.array([0, 0.5, 1, 1, 1, 1,1,1,1,1,1])) 
    focal_loss = sm.losses.BinaryFocalLoss() if n_classes == 1 else sm.losses.CategoricalFocalLoss()
    total_loss = dice_loss + (1 * focal_loss)

**Metrics:**

    metrics = [sm.metrics.IOUScore(threshold=0.5), sm.metrics.FScore(threshold=0.5)] 

**Model Compile and Train:**

    optim = keras.optimizers.Adam()
    
    model.compile(optim, total_loss, metrics)

After the model has been trained, I used the following function to evaluate the metrics on test set.

    test_dataset = Dataset(
        x_test_dir, 
        y_test_dir, 
        classes=CLASSES, 
        augmentation=get_validation_augmentation(),
        preprocessing=get_preprocessing(preprocess_input),
    )
    
    test_dataloader = Dataloder(test_dataset, batch_size=1, shuffle=False)
    
    
    
    
    scores = model.evaluate_generator(test_dataloader)
    
    print("Loss: {:.5}".format(scores[0]))
    for metric, value in zip(metrics, scores[1:]):
        print("mean {}: {:.5}".format(metric.__name__, value))

And it gives the following values.

> Loss: 0.84531 mean iou_score: 0.77155 mean f1-score: 0.80823

**Question:**
I have multiclass segmentation problem. Does this above `mean f1-score` refers to mean f1 score for all classes on all the test examples? How to calculate mean F1 score per class across all images?

**Try:**

I tried to compute F1 score (all classes) for each of 99 test images I have as follows.

    from sklearn.metrics import f1_score
    import statistics
    import warnings
    
    imagesList = []
    gtMasksList = []
    prMasksList = []
    scoresList = []
    
    for samples in (range(len(test_dataset))):
        
    
    
        image, gt_mask = test_dataset[samples]
        image = np.expand_dims(image, axis=0)
        pr_mask = model.predict(image)
    
        gt_mask = np.argmax(gt_mask, axis = 2)
        pr_mask = np.argmax(pr_mask[0], axis = 2)
    
        #score = classification_report(gt_mask.flatten(), pr_mask.flatten(), output_dict=True,  zero_division=1)
        
        score = f1_score(gt_mask.flatten(), pr_mask.flatten(), average='macro', zero_division=1)
        imagesList.append(image)
        gtMasksList.append(gt_mask)
        prMasksList.append(pr_mask)
        scoresList.append(score) 

Then I took a mean value of all F1 scores of all the images as follows.

    statistics.mean(scoresList)

Which gave me `0.8561783196548821` .

I want to compute per class F1 score?

  [1]: https://github.com/qubvel/segmentation_models/blob/master/examples/multiclass%20segmentation%20(camvid).ipynb

**Comment by tonyboston-au** [comment]
You could try pycm: https://pypi.org/project/pycm/ Supports calculation of F1 scores by class and many other statistics...
"PyCM is a multi-class confusion matrix library written in Python that supports both input data vectors and direct matrix, and a proper tool for post-classification model evaluation that supports most classes and overall statistics parameters."

**Comment by elliestath** [comment]
hey @tonyboston-au, actually the pycm library works very well, but I see different score results than the ones created in sm for the test set.
For example, I do calculate: 
```
scores = model.evaluate_generator(test_dataloader)

print("Loss: {:.4}".format(scores[0]))
for metric, value in zip(metrics, scores[1:]):
    print("mean {}: {:.4}".format(metric.__name__, value)) 
```

and I get:

```
Loss: 0.02967
mean iou_score: 0.6739
mean f1-score: 0.7038
```

However, when I use the pycm library I get F1 score per class and their average is `F1=0.7750`

Both time I do preprocessing with the same backbone and do not apply augmentation on the test set.
Any ideas?

**Comment by tonyboston-au** [comment]
Hi @elliestath 
I have noticed this as well and don't know why the differences occur. Sometimes pycm gives better stats (e.g. overall accuracy and mean F1) than sm and sometimes the reverse occurs. Might be a question for @qubvel?
